### ISSUE TITLE
ValueObject should implement IComparable

### ISSUE THREAD

**Issue by ahedreville** [vkhorikov/CSharpFunctionalExtensions#203]
Dear Vladimir

I am getting **System.ArgumentException : At least one object must implement IComparable**
using a _ValueObject in a OrderBy predicate_ when testing with ef core InMemory provider

notice that with SqlServer provider no exception issues and sorts.

It seems the implementation of the InMemory provider needs that the ValueObject should implement IComparable.

I added such interface to my Name ValueObject and solved for such case.
i.e. `public int CompareTo(object? obj) => Value.CompareTo((obj as Name)?.Value)`;

Could you please consider to add such interface to your library ValueObject?

Thanks in advance.

**Comment by vkhorikov** [comment]
Don't see why not. Feel free to submit a PR (this needs to be implemented in the base `ValueObject` class). Otherwise, I'll try to add this in the next week or two.

**Comment by ahedreville** [comment]
by now what I am doing is to derive from ValueObject.

```
    public abstract class BaseValueObject : ValueObject, IValueObject, IComparable
    {
        protected BaseValueObject()
        {
        }

        public abstract object GetValue();

        public abstract int CompareTo(object? obj);
    }

```
and from there..

```
    public class FullName : BaseValueObject
    {
        public string FirstName { get; }
        public string LastName { get; }

        protected FullName()
        : this("", "")
        {
        }

        protected FullName(string firstName, string lastName)
        {
            FirstName = firstName;
            LastName =  lastName;
        }

        public static Result<FullName> Create(string firstName, string lastName)
        {
            firstName = (firstName ?? "").TrimSpaces();
            lastName = (lastName ?? "").TrimSpaces();

            if (firstName.Length > 60) return Result.Failure<FullName>($"{nameof(firstName)} is too long");
            if (firstName.Length == 0) return Result.Failure<FullName>($"{nameof(firstName)} is empty");

            if (lastName.Length > 60) return Result.Failure<FullName>($"{nameof(lastName)} is too long");
            if (lastName.Length == 0) return Result.Failure<FullName>($"{nameof(lastName)} is empty");

            if (firstName.Equals(lastName, StringComparison.OrdinalIgnoreCase)) return Result.Failure<FullName>($"{nameof(firstName)} should not be equal to {nameof(lastName)}");

            return Result.Success(new FullName(firstName, lastName));
        }

        protected override IEnumerable<object> GetEqualityComponents()
        {
            yield return FirstName.ToLowerInvariant();
            yield return LastName.ToLowerInvariant();
        }

        public override object GetValue() => $"{FirstName} {LastName}";
        public override int CompareTo(object? obj) => string.Compare($"{FirstName} {LastName}", $"{(obj as FullName)?.FirstName} {(obj as FullName)?.LastName}", StringComparison.Ordinal);

        public static implicit operator string(FullName fullName) => $"{fullName.FirstName} {fullName.LastName}"; // every fullName is a string

    }

```
I also implemented a handy interface (not related to this discussion here, but nice to know about it) 

    public interface IValueObject 
    {
        object GetValue();
    }

used to help Newtonsoft.Json to convert the  ValueObject to json through the following json converter below

```
   // https://stackoverflow.com/questions/24472404/json-net-how-to-serialize-object-as-value

    public class ValueObjectConverter : JsonConverter
    {
        private static readonly ConcurrentDictionary<Type, Type> CtorArgsTypes = new ConcurrentDictionary<Type, Type>();

        public override void WriteJson(JsonWriter writer, object? value, JsonSerializer serializer)
        {
            if (!(value is IValueObject valueObject)) return;

            serializer.Serialize(writer, valueObject.GetValue());
        }

        public override object? ReadJson(JsonReader reader, Type objectType, object? existingValue, JsonSerializer serializer)
        {
            var parameterType = CtorArgsTypes.GetOrAdd(objectType, type => objectType.GetConstructors(BindingFlags.Public | BindingFlags.Instance).First().GetParameters().Single().ParameterType);

            var value = serializer.Deserialize(reader, parameterType);

            return Activator.CreateInstance(objectType, value);
        }

        public override bool CanConvert(Type objectType) => typeof(IValueObject).IsAssignableFrom(objectType);
    }

```

since adding interfaces could lead to breaking changes consider the new C# 8.0 Interfaces enhancements with default method implementation

BR

**Comment by ahedreville** [comment]
Just to recap a bit…

When dealing with value objects and EF Core collateral side effects arise.

EF Core has two mechanics to implement value objects: 

1) HasConversion:	Mainly used with single property value object

2) OwnsOne:		used with multiple properties value object

i.e. the below single property value object…

```
public class Name : ValueObject
{
    public string Value { get; }

    protected Name()
    : this("")
    {
    }

    protected Name(string value)
    {
        Value = value;
    }

    public static Result<Name> Create(string name)
    {
        name = (name ?? "").Trim();

        if (name.Length > 60) return Result.Failure<Name>($"{nameof(name)} is too long");
        if (name.Length == 0) return Result.Failure<Name>($"{nameof(name)} is empty");

        return Result.Success(new Name(name));
    }

    protected override IEnumerable<object> GetEqualityComponents()
    {
        yield return Value;
    }

    public static implicit operator string(Name name) => name.Value; // every name is a string
}

```
will be mapped on EF Core with…

`builder.Property(e => e.Name).HasConversion(e => e.Value, v => Name.Create(v).Value);`


and i.e. the below multiple properties value object

```
public class FullName : ValueObject
{
    public string FirstName { get; }
    public string LastName { get; }

    protected FullName()
    : this("", "")
    {
    }

    protected FullName(string firstName, string lastName)
    {
        FirstName = firstName;
        LastName =  lastName;
    }

    public static Result<FullName> Create(string firstName, string lastName)
    {
        firstName = (firstName ?? "").Trim();
        lastName = (lastName ?? "").Trim();

        if (firstName.Length > 60) return Result.Failure<FullName>($"{nameof(firstName)} is too long");
        if (firstName.Length == 0) return Result.Failure<FullName>($"{nameof(firstName)} is empty");

        if (lastName.Length > 60) return Result.Failure<FullName>($"{nameof(lastName)} is too long");
        if (lastName.Length == 0) return Result.Failure<FullName>($"{nameof(lastName)} is empty");

        if (firstName.Equals(lastName, StringComparison.OrdinalIgnoreCase)) return Result.Failure<FullName>($"{nameof(firstName)} should not be equal to {nameof(lastName)}");

        return Result.Success(new FullName(firstName, lastName));
    }

    protected override IEnumerable<object> GetEqualityComponents()
    {
        yield return FirstName;
        yield return LastName;
    }

    public static implicit operator string(FullName fullName) => $"{fullName.FirstName} {fullName.LastName}"; // every fullName is a string
}

```
will be mapped EF Core with…

```
builder.OwnsOne(e => e.FullName, o =>
{
    o.Property(e => e.FirstName).HasColumnName("FullNameFirstName");
    o.Property(e => e.LastName).HasColumnName("FullNameLastName");
});

```

However only the EF Core HasConversion implementation has some issues.

If you try to order as follows `OrderBy(d => d.Name.Value)`

Exception: **The LINQ expression OrderBy(d => d.Name.Value) could not be translated** is thrown.

Is not a not big problem we can just `OrderBy(d => d.Name)` what is more concise.

Other hidden issue is related to how in memory provider collections are implemented by underlying List and Lists Sort method.

Exception: **At least one object must implement IComparable** will occur when testing with in memory provider.

Furthermore, _I personally don’t like the verbosity_ that is rendered when converting to json the value object

Rather than this output format.

```
{
    "name": {
        "value": "John"
    },
    "fullName": {
        "firstName": "John",
        "lastName": "Rambo"
    },
    "email": {
        "value": "john@corp.com"
    },
    "dollar": {
        "amount": 10.0,
        "symbol": "USD"
    },
    "description": {
        "value": "IT Supporter"
    },
    "creditCard": {
        "expYear": 2020,
        "expMonth": 10,
        "number": "13133",
        "cardType": {
            "name": "Visa",
            "id": 0
        }
    },
    "nationalId": {
        "value": "312544NN"
    },
    "account": {
        "value": "124514420"
    },
    "comment": {
        "value": "That was the river, This is the sea!!!"
    },
    "direction": {
        "addressLine1": "4777 Rockne Drive",
        "addressLine2": "Landing Avenue",
        "city": "Bellevue",
        "postalCode": "98004"
    },
    "dateTimeRange": {
        "end": "2020-05-30T13:30:00.4832054+02:00",
        "start": "2020-05-29T13:30:00.4832054+02:00"
    },
    "phone": {
        "value": "610457895"
    }
}

```
I like to have this other output format

```
{
    "name": "John",
    "fullName": "John Rambo",
    "email": "john@corp.com",
    "dollar": "10 USD",
    "description": "IT Supporter",
    "creditCard": "Visa 13133 10 2020",
    "nationalId": "312544NN",
    "account": "124514420",
    "comment": "That was the river, This is the sea!!!",
    "direction": "4777 Rockne Drive Landing Avenue Bellevue 98004",
    "dateTimeRange": "30/05/2020 0:44:47 +02:00 29/05/2020 0:44:47 +02:00",
    "phone": "610457895"
}
```

One approach to deal with those issues is to **implement two interfaces** on the ValueObject class itself.

The **IComparable interface** requested by EF Core in memory provider.

And an additional custom interface, the **IValueObject interface**.

```
public interface IValueObject 
{
    object GetValue(string separator = " ");

    IEnumerable<object> GetEqualityComponents();
}
```

_The IValueObject interface, defines two methods_:

_The already known GetEqualityComponents_, and _a new GetValue method_.

We w**ill be using this new interface as a marker interface for a json value converter**
**and calling the new GetValue method to get the value that should be provided** when the conversion of the value object property happens.

To avoid breaking changes, we could provide overridable methods with default implementations on the ValueObject class itself.

Here below there are **simple implementations based on ToString()**

```
public abstract class ValueObject, IValueObject, IComparable
{
    default implementations......

    public virtual object GetValue(string separator = " ") => string.Join(separator, GetEqualityComponents().Select(c => c.ToString()));

#pragma warning disable CS8600 // Converting null literal or possible null value to non-nullable type.
    public virtual int CompareTo(object? obj) => string.CompareOrdinal(string.Join("", GetEqualityComponents().Select(c => c.ToString())), string.Join("", ((IEnumerable<object>)obj).Select(c => c.ToString())));
#pragma warning restore CS8600 // Converting null literal or possible null value to non-nullable type.
}
```


BR.

Tested with a mocking class like the one below with several properties value objects... (value object code not included in this comment for brevity, and the ValueObjectConverter class has been described on the previous comment)

```
var foo = new MyClass
{
    Name = Name.Create("John").Value,
    FullName = FullName.Create("John", "Rambo").Value,
    Email = Email.Create("John@corp.com").Value,
    Dollar = Dollar.Ten(),
    Description = Description.Create("IT Supporter").Value,
    CreditCard = CreditCard.Create("13133", 10, DateTimeOffset.UtcNow.Year, CreditCardType.Visa).Value,
    NationalId = NationalId.Create("312544NN").Value,
    Account = Account.Create("124514420").Value,
    Comment = Comment.Create("That was the river, This is the sea!!!").Value,
    Direction = Direction.Create("4777 Rockne Drive", "Landing Avenue", "Bellevue", "98004").Value,
    DateTimeRange = DateTimeRange.CreateOneDayRange(DateTimeOffset.Now).Value,
    Phone = Phone.Create("610457895").Value
};

```

**Comment by vkhorikov** [comment]
I've checked in the `IComparable` implementation.

Regarding the `GetValue` method, looks like it doesn't have to be an interface. The base `ValueObject` class will work too, correct?

**Comment by vkhorikov** [comment]
Closing this issue, the IComparable implementation is available in 2.12.0
